### DISCUSSION TITLE
Improving Readability of Code

### DISCUSSION THREAD

**Discussion by ghost** [h3rald/min#129] [Features: On Hold]
I thought a readability improvement idea, ex.:
`[5 6 -] [4 [1 2 +] *]`
We could use non-code elements(escapeable elements) like `[]` to tell users what is going on code.

**Comment by h3rald** [comment]
😳

Hmmm... I can't understand that, sorry... also > and < are used in (system) symbols and sigils

**Reply by ghost** [reply]
Now relook it, you'll see changes and hopefully you'll understand what i mean

**Reply by ghost** [reply]
> I thought a readability improvement idea, ex.:
> `[5 6 -] [4 [1 2 +] *]`
> We could use non-code elements(escapeable elements) like `[]` to tell users what is going on code.

At the end parser would escape them `[]`:
`[5 6 -] [4 [1 2 +] *] same as 5 6 - 4 1 2 + *`

So `[]` would be like... comments which are escaped by parser. But do you see blocks of code which explain the code:
`[5 6 -]` `[1 2 +]` `[4 [1 2 +] *]`

**Reply by h3rald** [reply]
Ahhh better, yes, you are basically adding square brackets to identify the values which will be consumed by a symbol.

The idea in itself is not bad, and yes, it would be very easy to update the parser to ignore square brackets on its own!

The only problem I have with it is that I wanted to keep square brackets for something else special, I just don't know what.

In TCL square brackets are used for evaluation -- I don't need them in min right now but it seems "a waste of a good symbol" to just ignore them... not sure what other characters we could use though, most of them are already used for sigils and system symbols.

**Reply by ghost** [reply]
Maybe we should use different symbols for the concept ??
`~~` `^^` etc.

**Reply by h3rald** [reply]
Yes definitely... ^ is for lambda and ~ will probably become lambda-bind. Ideally I'd use only one character, better if matching like brackets but I could live with just one character... which though...

Maybe `,` or `_`?

```
; Your proposal:
[5 6 -] [1 2 +] [4 [1 2 +] *]

; Another possibility:
,5 6 -, ,1 2 +, ,4 ,1 2 +, *,

; And another:
|5 6 -| |1 2 +| |4 |1 2 +| *|
```

Your proposal is definitely more readable. I am going to give it some serious thought. Also instead of not doing anything we could add additional checks, for example:

* `[`-- start recording 
* `5` -- one value added
* `6` -- another value added
* `-` -- symbol pushed on the stack
* `]` -- verify that all symbols from the last `[` have been consumed (*)


(*) Not sure how to implement this, because values have already been consumed by `-`.

I'd just like to use those symbols for something, you know?

**Reply by ghost** [reply]
> [-- start recording
> 5 -- one value added
> 6 -- another value added
> - -- symbol pushed on the stack
> ] -- verify that all symbols from the last [ have been consumed (*)

###Min###########
\# Achievement get!  #
\# More dicipline.      #
#############

> I'd just like to use those symbols for something, you know?

I didn't understand but if you want to use them at special areas then i may think what would be use cases of them.

**Reply by h3rald** [reply]
Yes, I would like them to do something, and something important... not getting ignored. Maybe I can find a way to monitor stack popping and keep track of what gets popped between a `[` and a `]`, handling nesting of course...

**Reply by h3rald** [reply]
I was thinking about this again and I am not sure: those brackets do complicate min code a lot. And one of the (weird) good things about concatenative languages is that you shouldn't need to use brackets or parenthesis to clarify precedence.

Putting this on hold for now.
